### PR TITLE
Add Code Actions, "Transform parameter objects into positional parameters" and vice-versa code action, to the Editor

### LINKED ISSUE #437: [Feature Request] [Editor] Action to transform a workshop action from using positional parameters to parameter objects, and vice versa
`Action(param1, param2, ...)` ↔ `Action({ Parameter 1: param1, Parameter 2: param2, ... })`

We could take a page out of VSCode's book and add a little 💡 icon with a list of actions depending on where the cursor is https://code.visualstudio.com/docs/editor/refactoring

### PR DESCRIPTION
**Pull request by netux**
## Code Actions

Similar to VSCode's code actions, add a lighbulb icon near the user's cursor that can be clicked to select refactor actions[1] in the code

Lightbulb
Lightbulb dropdown open

[1]: In reality, these are not limited to code refactors, but are meant for it anyway

## Transform parameter objects into positional parameters (and vice-versa) code action provider

As the name implies, they convert action arguments to be in one form or the other. E.g. `{ Param 1: value1, Param 2: value2 }` → `(value1, value2)`.

Takes into consideration parameter order and "autocomplete-min-parameter-newlines" for multi-line parameter objects

https://github.com/Mitcheljager/workshop.codes/assets/6181929/932eb9ce-b257-4087-8474-66a30eeb4263

Fixes #437

### CODE REVIEW

File: app/assets/stylesheets/elements/_codemirror.scss [hunk 1]
```diff
@@ -303,4 +303,27 @@
...
+  .cm-codeActionsLightbulb {
+    position: absolute;
+    aspect-ratio: 1;
+    line-height: 1;
+    font-size: 0.75rem;
+    background-color: $bg-darker;
+    padding: 0.25rem;
+    border-radius: 2px;
+    cursor: pointer;
+    transform: translate(-100%, -50%);
```
Comment (Mitcheljager) [inline app/assets/stylesheets/elements/_codemirror.scss:316]: Everywhere else we use the longer value
```suggestion
    transform: translateX(-100%) translateY(-50%);
```

File: app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+  const openParameterObjectBracketIndex = actionCallCont.slice(openParenIndex + 1).match(/(?<=^\s*)\{/)?.index ?? -1
+  if (openParameterObjectBracketIndex >= 0) {
+    // parameter object → positional parameters
+
+    const closingParameterObjectBracketIndex = getClosingBracket(actionCallCont, "{", "}", 0)
+    if (closingParameterObjectBracketIndex < 0) return
+
+    const fileWideParameterObjectFrom = (fileWideParenFrom + 1) + openParameterObjectBracketIndex
+    const fileWideParameterObjectTo = (fileWideParenFrom + 1) + closingParameterObjectBracketIndex
```
Comment (Mitcheljager) [inline app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js:47]: The brackets seem a little silly, but it's either or
```suggestion
    const fileWideParameterObjectFrom = fileWideParenFrom + openParameterObjectBracketIndex + 1
    const fileWideParameterObjectTo = fileWideParenFrom + closingParameterObjectBracketIndex + 1
```

File: app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    actions.push({
+      label: "Transform object into positional parameters",
```
Comment (Mitcheljager) [inline app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js:50]: Perhaps we can change the wording to "regular parameters" or "workshop syntax" as "positional parameters" is a bit technical

File: app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      position: fileWideActionNameStart,
+      run() {
+        const parameterObject = state.doc.sliceString(fileWideParameterObjectFrom + 1, fileWideParameterObjectTo - 1)
```
Comment (Mitcheljager) [inline app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js:53]: It seems a bit odd that this also does + 1, which is also added to the variable a couple lines before.  Is that correct?

File: app/javascript/src/lib/codeActions.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+class CodeActionLightBulbWidget extends WidgetType {
+  constructor(actions) {
+    super()
+
+    this.actions = actions
+  }
+
+  toDOM() {
+    const container = document.createElement("span")
+    container.className = "button cm-codeActionsLightbulb"
```
Comment (Mitcheljager) [inline app/javascript/src/lib/codeActions.js:23]: Since we can pick the classname here we might as well pick something nicer, even if it doesn't quite follow codemirror conventions

File: app/javascript/src/lib/codeActions.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    container.textContent = "💡"
+
+    const dropdown = document.createElement("select")
+    dropdown.className = "cm-codeActionsLightbulb__dropdown"
```
Comment (Mitcheljager) [inline app/javascript/src/lib/codeActions.js:27]: Same here

File: app/javascript/src/lib/codeActions.js [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+    this.actions = actions
+  }
+
+  toDOM() {
+    const container = document.createElement("span")
+    container.className = "button cm-codeActionsLightbulb"
+    container.textContent = "💡"
+
+    const dropdown = document.createElement("select")
```
Comment (Mitcheljager) [inline app/javascript/src/lib/codeActions.js:26]: Why is this a select with options rather than a more regular div with buttons inside?

File: app/javascript/src/lib/codeActions.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+        if (actions.length > 0) {
+          const { from } = update.state.selection.main
+          const decorationPosition = actions.reduce((position, action) => Math.min(action.position ?? Infinity, position), from)
+
+          this.decorations = Decoration.set([
+            createLightBulbDecoration(decorationPosition, actions)
+          ])
+        }
+      }
```
Comment (Mitcheljager) [inline app/javascript/src/lib/codeActions.js:91]: We could do with some fewer indents here

```suggestion
      if (!(update.docChanged || update.selectionSet || update.viewportChanged))
      
      this.decorations = Decoration.none

      const actions = providers
        .map((provider) => provider(update) ?? [])
        .flat()

      if (!actions.length) return
      
      const { from } = update.state.selection.main
      const decorationPosition = actions.reduce((position, action) => Math.min(action.position ?? Infinity, position), from)

      this.decorations = Decoration.set([
        createLightBulbDecoration(decorationPosition, actions)
      ])
```

File: app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+      run() {
+        const args = splitArgumentsString(state.doc.sliceString(fileWideParenFrom + 1, fileWideParenTo - 1))
+
+        const isMultiLine = args.length > get(settings)["autocomplete-min-parameter-newlines"]
+        const indent = "\t".repeat(getIndentForLine(state, cursorFrom, cursorFrom))
+
+        let parameterObject = "{"
+        if (isMultiLine) parameterObject += "\n"
+        else parameterObject += " "
+        parameterObject += args
```
Comment (Mitcheljager) [inline app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js:86]: Let;s add some new lines, it's a little hard to follow

```suggestion
        let parameterObject = "{"
        
        if (isMultiLine) parameterObject += "\n"
        else parameterObject += " "
        
        parameterObject += args
```

File: app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+          .map((value, index) => {
+            const keyValue = `${ completion.parameter_keys[index] }: ${ value.trim() }`
+            return (isMultiLine ? (indent + "\t") : "") + keyValue
+          })
+          .join(",\n")
+        if (isMultiLine) parameterObject += "\n" + indent
+        else parameterObject += " "
+        parameterObject += "}"
```
Comment (Mitcheljager) [inline app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js:94]:
```suggestion
        if (isMultiLine) parameterObject += "\n" + indent
        else parameterObject += " "
        
        parameterObject += "}"
```

**Review by Mitcheljager — CHANGES_REQUESTED** [review]
Very cool feature, nice use of widgets. I left quite a few comments, mostly concerning of formatting.

File: app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js (hunk 2 above)
Comment (netux) [reply app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js:47]: Thee brackets are there for readability. I wanted these calculations to read like "starting from the opening parenthesis, plus the index of the parameter object opening/closing bracket (which is already relative to that opening parenthesis)"

In my head it makes sense that we have to +1 the `fileWideParenFrom`, since we want to get _past_ that parenthesis.

File: app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js (hunk 3 above)
Comment (netux) [reply app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js:50]: I was hoping people would learn what positional parameters are along the way, but sure.

"Positional parameters" sounds weird anyway, since normally these would be called "positional arguments"[[1]]. But we already called the object notation "parameter objects" instead of "argument objects", so eh.

[1]: https://stackoverflow.com/a/156787

File: app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js (hunk 4 above)
Comment (netux) [reply app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js:53]: Yes, the +1 and -1 here are to get the stuff inside the curly braces (the "inner content" of the parameter objects).

File: app/javascript/src/lib/codeActions.js (hunk 5 above)
Comment (netux) [reply app/javascript/src/lib/codeActions.js:23]: I saw we already had some custom stuff using the `cm-` notation and just went with it as well.
But what would you suggest?

File: app/javascript/src/lib/codeActions.js (hunk 7 above)
Comment (netux) [reply app/javascript/src/lib/codeActions.js:26]: Easier to implement 😛. Also better for mobile, I guess.

If it looks too ugly I guess I can try and change it.

File: app/javascript/src/lib/codeActions.js (hunk 8 above)
Comment (netux) [reply app/javascript/src/lib/codeActions.js:91]: 👍. I left the stuff from the example here.

File: app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js (hunk 10 above)
Comment (netux) [reply app/javascript/src/lib/codeActionProviders/transformParameterObjectsIntoPositionalParameters.js:94]: Adding an extra newline for consistency

```suggestion

        if (isMultiLine) parameterObject += "\n" + indent
        else parameterObject += " "
        
        parameterObject += "}"
```

File: app/javascript/src/lib/codeActions.js (hunk 5 above)
Comment (Mitcheljager) [reply app/javascript/src/lib/codeActions.js:23]: I think it makes sense to use classnames of our own here and keep `cm-` to codemirror specific things so we can always differentiate between the two

File: app/javascript/src/lib/codeActions.js (hunk 7 above)
Comment (Mitcheljager) [reply app/javascript/src/lib/codeActions.js:26]: Fair enough!

File: app/javascript/src/lib/codeActions.js (hunk 5 above)
Comment (netux) [reply app/javascript/src/lib/codeActions.js:23]: Would it make sense to keep that in codeMirror.scss, or should I create a separate file for the extension?

Comment (Mitcheljager) [reply app/javascript/src/lib/codeActions.js:23]: Hmm good question, do what you think is best. That file is a bit messy unfortunately

Comment (netux) [reply app/javascript/src/lib/codeActions.js]: For now I'll just leave it in `_codemirror.scss`. Splitting that file up can be done in another PR :)

I also went with `.code-action-lightbulb` for the class name.

**Comment by netux** [comment]
Added a quick check to avoid providing the parameter object → positional parameters action if there is some more arguments after the parameter object

example

**Review by Mitcheljager — APPROVED** [review]
Nice!
